pandas/core/groupby/ops.py

```diff
@@ -260,6 +264,9 @@ def _get_output_shape(self, ngroups: int, values: np.ndarray) -> Shape:
     def _get_out_dtype(self, dtype: np.dtype) -> np.dtype:
         how = self.how
 
+        if self.kind == "transform" and self.grouper.has_dropped_na:
+            dtype = ensure_dtype_can_hold_na(dtype)
+
```

Comment:
ATM we do something similar to this L578-L591.  could this be rolled into that?  (admittedly thats a bit messy so this may just b A Better Solution)

Reply:
It seems better to me to determine the result dtype upfront, rather than cast after, as much as possible.

Reply:
i dont have an opinion on before vs after so am happy to defer to you, but do strongly prefer doing things Just One Way.

Reply:
Makes sense - I didn't realize that my change to `_get_cython_vals` actually makes `values` float, and so this is indeed not necessary at all.